# Convolutional Neural Networks

Convolutional neural networks, which is also called as convnet or CNN, is used for fomputer vision application

and have shown its outstanding performance, way better than fully connected layers. Let's talk about how we make simple example of CNN.

## 1. Basic CNN



In [6]:
from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

##### input shape

In CNN, it expect to have (image_height, image_width, image_channels) tensor for input, not like we used to do (n_x, ) with m.

##### model overview

In [7]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_2 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 64)                36928     
__________

This is typical form of CNN. (conv - pooling - conv - pooling - ... -flatten - dense - dense - softmax).

And networks are getting deeper and has smaller height and width.

##### Dataset (MNIST)

In [8]:
from keras.datasets import mnist
from keras.utils import to_categorical   # to one hot vector

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 109s 2ms/step - loss: 0.1644 - acc: 0.9480
Epoch 2/5
60000/60000 [==============================] - 93s 2ms/step - loss: 0.0479 - acc: 0.9852
Epoch 3/5
60000/60000 [==============================] - 107s 2ms/step - loss: 0.0327 - acc: 0.9896
Epoch 4/5
60000/60000 [==============================] - 104s 2ms/step - loss: 0.0254 - acc: 0.9920 0s - loss: 0.0255 - acc: 0
Epoch 5/5
60000/60000 [==============================] - ETA: 0s - loss: 0.0195 - acc: 0.993 - 100s 2ms/step - loss: 0.0195 - acc: 0.9939


##### Evaluate on testset

In [9]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
test_acc

10000/10000 [==============================] - 5s 539us/step


0.9902

99.02% shows better performance compared to fully connected networks'.(97.8%)